File: csharp/squeaky-clean/SqueakyClean.cs
using System;
using System.Text;
using System.Text.RegularExpressions;

public static class Identifier
{
    public static string Clean(string identifier)
    {
        var cleaned = new StringBuilder();
        var kebab = false;

        foreach (char c in identifier)
        {
            if (c == '-') {
                kebab = true;
                 // It will get filtered out by not being a letter later anyway
                continue;
            }

            if (kebab) {
                kebab = false;
                if (Char.IsLetter(c)) {
                    cleaned.Append(Char.ToUpper(c));
                    continue;
                }
            }
            
            if (Char.IsWhiteSpace(c)) {
                cleaned.Append('_');
                continue;
            }

            if (Char.IsControl(c)) {
                cleaned.Append("CTRL");
                continue;
            }

            if (!Char.IsLetter(c) || Regex.IsMatch(c.ToString(), @"[α-ω]")) {
                continue;
            }

            cleaned.Append(c);
        }

        return cleaned.ToString();
    }
}

File: java/squeaky-clean/src/main/java/SqueakyClean.java
class SqueakyClean {
    static String clean(String identifier) {
        var sb = new StringBuilder();
        char[] leetspeak = {'4', '3', '0', '1', '7'};
        char[] normspeak = {'a', 'e', 'o', 'l', 't'};

        for (int i = 0; i < identifier.length(); i++) {
            var c = identifier.charAt(i);
            if (c == ' ') {
                c = '_';
            }

            if (c == '-') {
                i++;
                if (i >= identifier.length()) {
                    break;
                }
                c = Character.toUpperCase(identifier.charAt(i));
            }

            for (int j = 0; j < 5 ; j++) {
                if (leetspeak[j] == c) {
                    c = normspeak[j];
                }
            }

            if (!Character.isLetter(c) && c != '_') {
                continue;
            }

            sb.append(c);
        }

        return sb.toString();
    }
}
